File: Codes/Work_Educate_Problem/Deterministic_nonlinear.jl
############## Deterministic optimization ##############

using JuMP, Gurobi, Plots

### Parameters 
# Numbers of periods
n = 20
# Base salary
base_salary = 1
# Base Education
base_education = 0
# Base Money
base_money = 3
# Expenses
expenses = 1
# Education Rate
education_rate = 2

model = Model(Gurobi.Optimizer)

@variable(model, w[1:n], Bin)
@variable(model, s[1:n], Bin)
@variable(model, e[1:n])
@variable(model, m[1:n])
@variable(model, p[1:n])

@objective(model, Max, sum(p[i] - expenses for i = 1:n))

# Payments constraint
@constraint(model, [i = 2:n], p[i] == w[i-1] * (base_salary * (1 + (e[i-1] / 2))))
@constraint(model, p[1] == 0)
# Education level
@constraint(model, [i = 2:n], e[i] == e[i-1] + s[i-1] * education_rate)
@constraint(model, e[1] == 0)
# Money constraint
@constraint(model, [i = 2:n], m[i] == m[i-1] + p[i] - expenses)
@constraint(model, m[1] == base_money)
@constraint(model, [i = 1:n], m[i] >= 0)
# Only one choice 
@constraint(model, [i = 1:n], w[i] + s[i] <= 1)

JuMP.optimize!(model)

work = value.(w)
study = value.(s)
money = value.(m)
education = value.(e)

println("Objective is: ", JuMP.objective_value(model))
println("Solution is: ")
println(work)
println(study)
println(money)
println(education)

# Time (x-axis)
time = 1:n

# Create the plot
p = plot(time, work, label="Work", marker=:circle, xlabel="Time", ylabel="Value", legend=:top)
plot!(p, time, study, label="Study", marker=:star5)
plot!(p, time, money, label="Money", marker=:diamond)
plot!(p, time, education, label="Education", marker=:cross)

# Display the plot
display(p)

### Comments ###
# Non linearity constraint in the constraint of the transition function

File: Codes/Sequential_Decisions/Théodore/Deterministic_nonlinear.py
############## Deterministic optimization ##############

import gurobipy as gb
from gurobipy import GRB
import matplotlib.pyplot as plt

### Parameters 
# Numbers of periods
n = 20
# Base salary
base_salary = 1
# Base Education
base_education = 0
# Base Money
base_money = 3
# Expenses
expenses = 1
# Education Rate
education_rate = 2

# Define the model
model = gb.Model("work_study_model") 
    
# Set time limit
model.Params.TimeLimit = 100

# Add variables 
work = {i: model.addVar(vtype = GRB.BINARY, name="Working at time {0}".format(i)) for i in range(n)}
study = {i: model.addVar(vtype = GRB.BINARY, name="Study at time {0}".format(i)) for i in range(n)}
education = {i: model.addVar(vtype = GRB.CONTINUOUS, lb=0, ub=GRB.INFINITY, name="Education level at time {0}".format(i)) for i in range(n)}
money = {i: model.addVar(vtype = GRB.CONTINUOUS, lb=0, ub=GRB.INFINITY, name="Money level at time {0}".format(i)) for i in range(n)}
profit = {i: model.addVar(vtype = GRB.CONTINUOUS, lb=0, ub=GRB.INFINITY, name="Profit at time {0}".format(i)) for i in range(n)}

# Set objective function
objective = gb.quicksum(profit[i] - expenses for i in range(n))
model.setObjective(objective, GRB.MAXIMIZE)

# Constraints
salary_constraint = {i : model.addConstr(
    profit[i],
    gb.GRB.EQUAL,
    work[i] * (base_salary * (1 + (education[i] / 2))),
    name="Profit definition at time {0}".format(i)
    ) for i in range(n)}

Education_constraint = {i : model.addConstr(
    education[i],
    gb.GRB.EQUAL,
    education[i-1] + study[i-1] * education_rate,
    name="Education level definition at time {0}".format(i)
    ) for i in range(1,n)}
Education_constraint[0] = model.addConstr(education[0],gb.GRB.EQUAL,base_education,name="Education level definition at time 0")

Money_constraint = {i : model.addConstr(
    money[i],
    gb.GRB.EQUAL,
    money[i-1] + profit[i-1] - expenses,
    name="Money level definition at time {0}".format(i)
    ) for i in range(1,n)}
Money_constraint[0] = model.addConstr(money[0],gb.GRB.EQUAL,base_money,name="Money level definition at time 0")

Choice_constraint = {i : model.addConstr(
    work[i] + study[i],
    gb.GRB.LESS_EQUAL,
    1,
    name="Choice definition at time {0}".format(i)
    ) for i in range(n)}

# Optimization
model.optimize()

for v in model.getVars():
    print('%s %g' % (v.VarName, v.X))

print('Objective function: %g' % model.ObjVal)

work = [work[i].x for i in range(n)]
study = [study[i].x for i in range(n)]
money = [money[i].x for i in range(n)]
education = [education[i].x for i in range(n)]

print(work)
print(study)
print(money)
print(education)

# Time (x-axis)
time = list(range(n))

# Create the plot
plt.plot(time, work, label="Work")
plt.plot(time, study, label="Study")
plt.plot(time, money, label="Money")
plt.plot(time, education, label="Education")

# Display the plot
plt.xlabel('Time')
plt.ylabel('Value')
plt.legend()
plt.show()

### Comments ###
# Non linearity constraint in the constraint of the transition function